kuma/wiki/management/commands/render_document.py

```diff
@@ -2,100 +2,115 @@
     )
 
     def handle(self, *args, **options):
         self.options = options
         self.base_url = options['baseurl'] or absolutify('')
+        if self.options['nocache']:
+            self.cache_control = 'no-cache'
+        else:
+            self.cache_control = 'max-age=0'
 
         if options['all']:
-            logging.info(u"Querying ALL %s documents..." %
-                         Document.objects.count())
-            docs = Document.objects.order_by('-modified').iterator()
-            for doc in docs:
-                self.do_render(doc)
+            # Query all documents, excluding those whose `last_rendered_at` is
+            # within `min_render_age` or NULL.
+            min_render_age = (
+                datetime.datetime.now() -
+                datetime.timedelta(seconds=self.options['min_age']))
+            docs = Document.objects.filter(
+                Q(last_rendered_at__isnull=True) |
+                Q(last_rendered_at__gte=min_render_age))
+            docs = docs.order_by('-modified')
+            self.chain_render_docs(docs)
 
         else:
             if not len(args) == 1:
-                raise CommandError("Need at least one document path to render")
+                raise CommandError('Need at least one document path to render')
             for path in args:
-                # Accept a single page path from command line, but be liberal in
-                # what we accept, eg: /en-US/docs/CSS (full path); /en-US/CSS (no
-                # /docs); or even en-US/CSS (no leading slash)
+                # Accept a single page path from command line, but be liberal
+                # in what we accept, eg: /en-US/docs/CSS (full path);
+                # /en-US/CSS (no /docs); or even en-US/CSS (no leading slash)
                 if path.startswith('/'):
                     path = path[1:]
                 locale, sep, slug = path.partition('/')
                 head, sep, tail = slug.partition('/')
                 if head == 'docs':
                     slug = tail
-                self.do_render(Document.objects.get(locale=locale, slug=slug))
-
-    def do_render(self, doc):
-        # Skip very recently rendered documents. This should help make it
-        # easier to start and stop an --all command without needing to start
-        # from the top of the list every time.
-        if doc.last_rendered_at:
-            now = datetime.datetime.now()
-            render_age = now - doc.last_rendered_at
-            min_age = datetime.timedelta(seconds=self.options['min_age'])
-            if (render_age < min_age):
-                logging.debug(u"Skipping %s (%s) - rendered %s sec ago" %
-                              (doc, doc.get_absolute_url(), render_age))
-                return
-
-        if self.options['force']:
-            doc.render_started_at = None
+                doc = Document.objects.get(locale=locale, slug=slug)
```

Comment:
Could we factor out that query into a command method so that we don't have to do a query per passed arg? Something like `get_documents_for_args` or something?